Saving places

PowerfulBacon opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds in a build mode tool that lets you save a selected area into the .dmm format.

Saving the map works great and now stores the right variables with the ability to add more (inspired by https://github.com/BeeStation/BeeStation-Hornet/pull/1526)

Alright, this is a map saver that converts areas of turfs into a string in TGM format.
This also adds in a proc to /obj which allows you to add text to the file when the object is saved, and is currently used to save the contents of lockers, storage containers and to spawn the contents of ORMs on top, which is better than them getting cleared.

This implementation is very flexible and has a few input options such as:
 - Save objects (Whether or not to save /obj)
 - Save mobs (Whether or not to save /mob)
 - Save turfs (Whether or not to save /turf (If not saves it as a NOOP turf))
 - Save areas (Whether or not to save /area (If not saves it as a NOOP area))
 - Save space (If not will be saved as space, useful for saving shuttles)
 - Save object properties (Whether or not to run the objects on_object_saved() proc, which can give extra text to add to the function)
 - Save shuttles areas only (And area that isn't a shuttle will be saved as noop)
 - save only non shuttles (Any area that is a shuttle will be saved as noop)

This also includes a map generator in build mode map generator, which will save all areas. It has been made easier to use but doesn't fully use the customisation that has been made possible.
When using FTP to send the file to the client, the map is saved on the server under data/temp.dmm and then transmitted. The maps are saved here before being transmitted, but the map is only saved on the server until another map is saved, and then the file is overwritten to prevent the server from being filled up with tons of files that the server owner doesn't care about.
This is also useful for coders who want to make something.

I was thinking something like:
 - A space ruin that will autosave and overwrite itself at the end of every game so is persistent
 - Admin events using persistent bases (hence the ORM, locker and storage implementation)

Flaws with my implementation:
 - Objects that are auto-filled (Toolboxes that spawn with things in) will save their contents, however when spawned in, their old contents will be saved and they will spawn with new default contents causing double of what they have, but this is worth it for saving items that are in lockers and bags since it could be frustrating to accidentally lose stuff.

Oh yea, it also saves into the TGM format, which is nice because it the format currently used to help prevent merge conflicts, so yea that's pretty nice :).

Put some examples of the files created with this in the comments blow (Newer and more functional ones towards the bottom)

## Why It's Good For The Game

Very, very useful for admins, allows you to save things to another round (Build part of a station and round ends, just save it in buildmode and continue another time)

## Changelog
:cl:
server: Adds in saving areas
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

PowerfulBacon committed:
holy fuck it worked.

PowerfulBacon committed:
Added full list of characters to the pool and removed the save tool (since its a map gen)

PowerfulBacon committed:
Generate Metadata proc

PowerfulBacon commented:
As I was writing this I got pinged with https://github.com/WaspStation/WaspStation-1.0/pull/109/files so I am going to steal some thing with the vars from there

PowerfulBacon commented:
Currently saving is fine, but loading the maps causes a few runtimes (could be the map loader rather than this, but I'll look into it anyway)

PowerfulBacon commented:
Also should be noted that this one doesn't compress the files down

PowerfulBacon committed:
Removed a warning, added in dynamic saving vars

PowerfulBacon commented:
Also this will make a new file on the server /data/temp.dmm when a level is saved, but this file is sent to clients and will be replaced by newer saves so won't eat up tons of data

PowerfulBacon commented:
This is an image of me copying part of the station
![image](https://user-images.githubusercontent.com/26465327/79079088-d7acf900-7d04-11ea-82e6-e5e362efc7b7.png)

As well as the file created
````
"ab" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"ac" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"ad" = (/obj/structure/sign/warning/electricshock{dir = 2;name = "HIGH VOLTAGE"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"ae" = (/obj/structure/cable{dir = 2;name = "power cable"},/obj/structure/grille{dir = 2;name = "grille"},/obj/structure/window/reinforced/fulltile{dir = 5;name = "reinforced window"},/turf/open/floor/plating,/area/ai_monitored/storage/eva)
"af" = (/obj/machinery/door/firedoor{dir = 2;name = "firelock"},/obj/machinery/door/airlock/command/glass{dir = 2;name = "EVA Storage"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 1;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 1;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"ag" = (/obj/structure/cable{dir = 2;name = "power cable"},/obj/structure/grille{dir = 2;name = "grille"},/obj/structure/window/reinforced/fulltile{dir = 5;name = "reinforced window"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plating,/area/ai_monitored/storage/eva)
"ah" = (/obj/structure/sign/warning/securearea{dir = 2;name = "SECURE AREA"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"ai" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"aj" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"ak" = (/obj/machinery/suit_storage_unit/standard_unit{dir = 2;name = "suit storage unit"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"al" = (/obj/machinery/suit_storage_unit/standard_unit{dir = 2;name = "suit storage unit"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"am" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"an" = (/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 1;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 1;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"ao" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"ap" = (/obj/machinery/suit_storage_unit/standard_unit{dir = 2;name = "suit storage unit"},/obj/machinery/light{dir = 2;name = "light fixture"},/obj/machinery/camera/autoname{dir = 1;name = "security camera"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aq" = (/obj/machinery/suit_storage_unit/standard_unit{dir = 2;name = "suit storage unit"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"ar" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"as" = (/obj/machinery/requests_console{dir = 2;name = "EVA requests console";pixel_x = "-32"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"at" = (/obj/machinery/atmospherics/components/unary/vent_pump/on/layer1{dir = 4;name = "EVA Storage vent pump #1"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"au" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"av" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/manifold/supply/hidden/layer1{dir = 4;name = "air supply pipe"},/obj/machinery/atmospherics/pipe/manifold/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/item/stack/telecrystal{dir = 2;name = "telecrystal";amount = "34"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aw" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/item/pen{dir = 2;name = "astronaut pen"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"ax" = (/obj/machinery/atmospherics/components/unary/vent_scrubber/on/layer3{dir = 8;name = "EVA Storage air scrubber #2"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"ay" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"az" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"aA" = (/obj/structure/rack{dir = 2;name = "rack"},/obj/machinery/light{dir = 8;name = "light fixture"},/obj/item/tank/jetpack/carbondioxide{dir = 2;name = "jetpack (carbon dioxide)"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aB" = (/obj/structure/rack{dir = 2;name = "rack"},/obj/item/tank/jetpack/carbondioxide{dir = 2;name = "jetpack (carbon dioxide)"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aC" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aD" = (/obj/machinery/holopad{dir = 2;name = "holopad"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 1;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 1;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aE" = (/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aF" = (/obj/structure/rack{dir = 2;name = "rack"},/obj/item/clothing/shoes/magboots{dir = 2;name = "magboots"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aG" = (/obj/structure/rack{dir = 2;name = "rack"},/obj/item/clothing/shoes/magboots{dir = 2;name = "magboots"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aH" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/door/firedoor{dir = 2;name = "firelock"},/obj/machinery/door/airlock/maintenance{dir = 2;name = "EVA Maintenance"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/maintenance/fore)
"aI" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aJ" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aK" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aL" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/manifold/scrubbers/hidden/layer3{dir = 1;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/manifold/supply/hidden/layer1{dir = 1;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aM" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aN" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aO" = (/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{dir = 8;name = "scrubbers pipe"},/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{dir = 8;name = "air supply pipe"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aP" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"aQ" = (/obj/structure/closet/crate/rcd{dir = 2;name = "RCD crate"},/obj/machinery/camera/autoname{dir = 2;name = "security camera"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel/dark,/area/ai_monitored/storage/eva)
"aR" = (/obj/structure/rack{dir = 2;name = "rack"},/obj/machinery/light{dir = 1;name = "light fixture"},/obj/item/hand_labeler{dir = 2;name = "hand labeler"},/obj/item/flashlight{dir = 2;name = "flashlight"},/obj/item/flashlight{dir = 2;name = "flashlight"},/obj/item/flashlight{dir = 2;name = "flashlight"},/obj/item/flashlight{dir = 2;name = "flashlight"},/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/open/floor/plasteel/dark,/area/ai_monitored/storage/eva)
"aS" = (/obj/machinery/firealarm{dir = 2;name = "fire alarm";pixel_y = "24"},/obj/item/clothing/head/welding{dir = 2;name = "welding helmet"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aT" = (/obj/machinery/power/apc/auto_name/north{dir = 2;name = "EVA Storage APC";pixel_y = "23"},/obj/structure/cable/yellow{dir = 2;name = "power cable"},/obj/machinery/power/terminal{dir = 1;name = "terminal"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aU" = (/obj/machinery/airalarm{dir = 2;name = "EVA Storage Air Alarm";pixel_y = "23"},/obj/item/radio/off{dir = 2;name = "station bounced radio"},/obj/item/assembly/timer{dir = 2;name = "timer"},/turf/open/floor/plasteel,/area/ai_monitored/storage/eva)
"aV" = (/obj/structure/table{dir = 2;name = "table"},/obj/item/stack/cable_coil{dir = 2;name = "cable coil";pixel_x = "-2";pixel_y = "-1";amount = "30"},/obj/machinery/cell_charger{dir = 2;name = "cell charger"},/obj/item/stock_parts/cell/high/plus{dir = 2;name = "high-capacity power cell+";pixel_x = "4";pixel_y = "-1"},/turf/open/floor/plasteel/dark,/area/ai_monitored/storage/eva)
"aW" = (/obj/structure/table{dir = 2;name = "table"},/obj/item/storage/toolbox/electrical{dir = 2;name = "electrical toolbox";pixel_x = "1";pixel_y = "-1"},/obj/item/screwdriver{dir = 2;name = "screwdriver";pixel_y = "16"},/turf/open/floor/plasteel/dark,/area/ai_monitored/storage/eva)
"aX" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"aY" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"aZ" = (/obj/effect/abstract/proximity_checker{dir = 2;name = "proximity checker"},/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"ba" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"bb" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"bc" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"bd" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)
"be" = (/turf/closed/wall/r_wall,/area/ai_monitored/storage/eva)



(1,1,1) = {"abacadaeafagahai"}
(1,2,1) = {"ajakalamanaoapaq"}
(1,3,1) = {"arasatauavawaxay"}
(1,4,1) = {"azaAaBaCaDaEaFaG"}
(1,5,1) = {"aHaIaJaKaLaMaNaO"}
(1,6,1) = {"aPaQaRaSaTaUaVaW"}
(1,7,1) = {"aXaYaZbabbbcbdbe"}`
```

PowerfulBacon commented:
Logic error thats causing values to not be saved at all, might have to continue this in morning

PowerfulBacon commented:
(value == initial(O.vars[V])) all the time
will fix tomorrow

ike709 commented:
Talk to mark, he added this to wasp already.

ike709 commented:
...see #1526

PowerfulBacon commented:
> ...see #1526

Yea got pinged with that on discord as I was writing the pr

PowerfulBacon committed:
Fixed a dumb dumb logic error that made it so only icons, files and text got saved. Now everything does

PowerfulBacon commented:
Ok it works now

PowerfulBacon committed:
Fixed not all keys getting used

PowerfulBacon committed:
Fixed

MCterra10 commented:
We implemented this downstream a while ago (I can't recall if it survived the rebase), but we used the standard dmm tools ported from Oracle. It might be a better solution, up to you

Edit: there's already a pr open that uses map exporter to facilitate saves

PowerfulBacon commented:
Started to continue work on this actually
Todo:
 - <s>TGM file format</s>
 - fdel(file) doesn't delete it from the server
 - <s>Save more vars (specifically contents is important) (So saving vars that are lists)<s>
 - Save decals
 - <s>Save Options (Ignore turfs, ignore mobs, ignore areas etc.)</s>
 - <s>More modularity (Have a proc that you give in the list of turfs and it will save it)</s>

PowerfulBacon commented:
Files are now exported in the TGM format.
```"a" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"b" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"c" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"d" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"e" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"f" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"g" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"h" = (
/obj/machinery/computer/bank_machine{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	light_power = 2;
	light_range = 2;
	obj_integrity = 200
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"i" = (
/obj/machinery/atmospherics/components/unary/vent_scrubber/on/layer3{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	id_tag = "34";
	uid = "34";
	radio_filter_out = "to_airalarm";
	radio_filter_in = "from_airalarm";
	initialize_directions = 4;
	obj_integrity = 200;
	name = "Vault air scrubber #1";
	icon_state = "scrub_on";
	dir = 4;
	plane = -2
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"j" = (
/obj/structure/closet/crate/goldcrate{
	flags_1 = 16384;
	obj_integrity = 200
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"k" = (
/obj/machinery/ore_silo{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 200
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"l" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"m" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"n" = (
/obj/machinery/light_switch{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 200;
	pixel_y = 28
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"o" = (
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	initialize_directions = 10;
	obj_integrity = 200;
	dir = 10;
	layer = 2.4;
	invisibility = 100;
	alpha = 64
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"p" = (
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	initialize_directions = 5;
	obj_integrity = 200;
	dir = 5;
	layer = 2.4;
	invisibility = 100;
	alpha = 64
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"q" = (
/obj/machinery/light{
	datum_flags = 4;
	flags_1 = 16384;
	on = 1;
	signal_enabled = 1;
	light_power = 0.75;
	light_range = 11;
	light_color = "#FFF6ED";
	switchcount = 1;
	active_power_usage = 79.2;
	on_gs = 1;
	static_power_used = 158.4;
	obj_integrity = 100
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"r" = (
/obj/structure/sign/warning/securearea{
	flags_1 = 16384;
	obj_integrity = 100
	},
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"s" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"t" = (
/obj/machinery/airalarm{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 250;
	name = "Vault Air Alarm";
	pixel_y = 23
	},
/obj/machinery/light{
	datum_flags = 4;
	flags_1 = 16384;
	on = 1;
	signal_enabled = 1;
	light_power = 0.75;
	light_range = 11;
	light_color = "#FFF6ED";
	switchcount = 1;
	active_power_usage = 79.2;
	on_gs = 1;
	static_power_used = 158.4;
	obj_integrity = 100;
	dir = 1
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"u" = (
/obj/machinery/nuclearbomb/selfdestruct{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	previous_level = "green";
	obj_integrity = 200
	},
/obj/effect/countdown/nuclearbomb{
	flags_1 = 16384;
	inertia_next_move = 50.5;
	obj_integrity = 500;
	color = "#81ff14"
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"v" = (
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	initialize_directions = 10;
	obj_integrity = 200;
	dir = 10;
	layer = 2.4;
	invisibility = 100;
	alpha = 64
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	initialize_directions = 6;
	obj_integrity = 200;
	dir = 6;
	layer = 2.3;
	invisibility = 100;
	alpha = 64
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"w" = (
/obj/structure/cable/yellow{
	flags_1 = 16384;
	cable_color = "#ffff00";
	d1 = 2;
	d2 = 4;
	obj_integrity = 300;
	icon_state = "2-4";
	invisibility = 100
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 200;
	dir = 1;
	layer = 2.3;
	invisibility = 100;
	alpha = 64
	},
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 200;
	dir = 1;
	layer = 2.4;
	invisibility = 100;
	alpha = 64
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"x" = (
/obj/machinery/door/airlock/vault{
	datum_flags = 4;
	flags_1 = 18432;
	locked = 1;
	signal_enabled = 1;
	req_access_txt = "53";
	explosion_block = -1;
	real_explosion_block = 2;
	damage_deflection = 30;
	obj_integrity = 600;
	max_integrity = 600;
	layer = 3.1
	},
/obj/structure/cable/yellow{
	flags_1 = 16384;
	cable_color = "#ffff00";
	d1 = 1;
	d2 = 2;
	obj_integrity = 300;
	icon_state = "1-2";
	invisibility = 100
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 200;
	dir = 1;
	layer = 2.3;
	invisibility = 100;
	alpha = 64
	},
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 200;
	dir = 1;
	layer = 2.4;
	invisibility = 100;
	alpha = 64
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"y" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"z" = (
/obj/machinery/power/apc/auto_name/north{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	light_range = 2;
	light_color = "#FA8282";
	tdir = 1;
	has_electronics = 2;
	update_state = 129;
	update_overlay = 12289;
	obj_integrity = 200;
	name = "Vault APC";
	dir = 2
	},
/obj/structure/cable/yellow{
	flags_1 = 16384;
	cable_color = "#ffff00";
	d2 = 2;
	obj_integrity = 300;
	icon_state = "0-2";
	invisibility = 100
	},
/obj/machinery/power/terminal{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	obj_integrity = 200;
	icon_state = "term-f";
	dir = 1;
	invisibility = 100
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"A" = (
/obj/structure/cable/yellow{
	flags_1 = 16384;
	cable_color = "#ffff00";
	d1 = 1;
	d2 = 2;
	obj_integrity = 300;
	icon_state = "1-2";
	invisibility = 100
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	initialize_directions = 6;
	obj_integrity = 200;
	dir = 6;
	layer = 2.3;
	invisibility = 100;
	alpha = 64
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"B" = (
/obj/structure/cable/yellow{
	flags_1 = 16384;
	cable_color = "#ffff00";
	d1 = 1;
	d2 = 2;
	obj_integrity = 300;
	icon_state = "1-2";
	invisibility = 100
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	volume = 70;
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	initialize_directions = 9;
	obj_integrity = 200;
	dir = 9;
	layer = 2.3;
	invisibility = 100;
	alpha = 64
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"C" = (
/obj/structure/cable/yellow{
	flags_1 = 16384;
	cable_color = "#ffff00";
	d1 = 1;
	d2 = 8;
	obj_integrity = 300;
	icon_state = "1-8";
	invisibility = 100
	},
/obj/machinery/light{
	datum_flags = 4;
	flags_1 = 16384;
	on = 1;
	signal_enabled = 1;
	light_power = 0.75;
	light_range = 11;
	light_color = "#FFF6ED";
	switchcount = 1;
	active_power_usage = 79.2;
	on_gs = 1;
	static_power_used = 158.4;
	obj_integrity = 100
	},
/obj/machinery/camera/autoname{
	datum_flags = 4;
	flags_1 = 16384;
	number = 1;
	signal_enabled = 1;
	c_tag = "Vault #1";
	obj_integrity = 100;
	dir = 1
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"D" = (
/obj/structure/sign/warning/securearea{
	flags_1 = 16384;
	obj_integrity = 100
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"E" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"F" = (
/obj/structure/filingcabinet{
	flags_1 = 16384;
	obj_integrity = 300
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"G" = (
/obj/machinery/atmospherics/components/unary/vent_pump/on/layer1{
	datum_flags = 4;
	flags_1 = 16384;
	signal_enabled = 1;
	id_tag = "39";
	uid = "39";
	radio_filter_out = "to_airalarm";
	radio_filter_in = "from_airalarm";
	initialize_directions = 8;
	obj_integrity = 200;
	name = "Vault vent pump #1";
	icon_state = "vent_out";
	dir = 8;
	plane = -2
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"H" = (
/obj/structure/closet/crate/silvercrate{
	flags_1 = 16384;
	obj_integrity = 200
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"I" = (
/obj/structure/safe{
	space = 10;
	flags_1 = 16384;
	tumbler_1_pos = 3;
	tumbler_1_open = 54;
	tumbler_2_pos = 56;
	tumbler_2_open = 7;
	obj_integrity = 300
	},
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"J" = (
/obj/effect/abstract/proximity_checker{
	flags_1 = 16384;
	obj_integrity = 500
	},
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"K" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"L" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"M" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"N" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"O" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"P" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)

(1,1,1) = {"
a
b
c
d
e
f
"}
(2,1,1) = {"
g
h
i
j
k
l
"}
(3,1,1) = {"
m
n
o
p
q
r
"}
(4,1,1) = {"
s
t
u
v
w
x
"}
(5,1,1) = {"
y
z
A
B
C
D
"}
(6,1,1) = {"
E
F
G
H
I
J
"}
(7,1,1) = {"
K
L
M
N
O
P
"}
```

PowerfulBacon committed:
Exporting in TGM format

PowerfulBacon committed:
Files are now more modular, as well as actually compressed

PowerfulBacon commented:
**Compression horray**
```
"a" = (
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"h" = (
/obj/machinery/computer/bank_machine,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"i" = (
/obj/machinery/atmospherics/components/unary/vent_scrubber/on/layer3{
	name = "Vault air scrubber #1";
	icon_state = "scrub_on";
	dir = 4
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"j" = (
/obj/structure/closet/crate/goldcrate,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"k" = (
/obj/machinery/ore_silo,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"n" = (
/obj/machinery/light_switch{
	pixel_y = 28
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"o" = (
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	dir = 10
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"p" = (
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	dir = 5
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"q" = (
/obj/machinery/light,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"r" = (
/obj/structure/sign/warning/securearea,
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"t" = (
/obj/machinery/airalarm{
	name = "Vault Air Alarm";
	pixel_y = 23
	},
/obj/machinery/light{
	dir = 1
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"u" = (
/obj/machinery/nuclearbomb/selfdestruct,
/obj/effect/countdown/nuclearbomb{
	color = "#81ff14"
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"v" = (
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	dir = 10
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	dir = 6
	},
/obj/effect/abstract/proximity_checker,
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"w" = (
/obj/structure/cable/yellow{
	icon_state = "2-4"
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	dir = 1
	},
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	dir = 1
	},
/obj/effect/abstract/proximity_checker,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"x" = (
/obj/machinery/door/airlock/vault{
	req_access_txt = "53"
	},
/obj/structure/cable/yellow{
	icon_state = "1-2"
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	dir = 1
	},
/obj/machinery/atmospherics/pipe/simple/scrubbers/hidden/layer3{
	dir = 1
	},
/obj/effect/abstract/proximity_checker,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"z" = (
/obj/machinery/power/apc/auto_name/north{
	name = "Vault APC";
	dir = 2
	},
/obj/structure/cable/yellow{
	icon_state = "0-2"
	},
/obj/machinery/power/terminal{
	icon_state = "term-f";
	dir = 1
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"A" = (
/obj/structure/cable/yellow{
	icon_state = "1-2"
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	dir = 6
	},
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"B" = (
/obj/structure/cable/yellow{
	icon_state = "1-2"
	},
/obj/machinery/atmospherics/pipe/simple/supply/hidden/layer1{
	dir = 9
	},
/obj/effect/abstract/proximity_checker,
/turf/open/floor/circuit,
/area/ai_monitored/nuke_storage)
"C" = (
/obj/structure/cable/yellow{
	icon_state = "1-8"
	},
/obj/machinery/light,
/obj/machinery/camera/autoname{
	dir = 1
	},
/obj/effect/abstract/proximity_checker,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"D" = (
/obj/structure/sign/warning/securearea,
/obj/effect/abstract/proximity_checker,
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)
"F" = (
/obj/structure/filingcabinet,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"G" = (
/obj/machinery/atmospherics/components/unary/vent_pump/on/layer1{
	name = "Vault vent pump #1";
	icon_state = "vent_out";
	dir = 8
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"H" = (
/obj/structure/closet/crate/silvercrate,
/obj/effect/abstract/proximity_checker,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"I" = (
/obj/structure/safe,
/obj/effect/abstract/proximity_checker,
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
"J" = (
/obj/effect/abstract/proximity_checker,
/turf/closed/wall/r_wall,
/area/ai_monitored/nuke_storage)

(1,1,1) = {"
a
a
a
a
a
a
"}
(2,1,1) = {"
a
h
i
j
k
a
"}
(3,1,1) = {"
a
n
o
p
q
r
"}
(4,1,1) = {"
a
t
u
v
w
x
"}
(5,1,1) = {"
a
z
A
B
C
D
"}
(6,1,1) = {"
a
F
G
H
I
J
"}
(7,1,1) = {"
a
a
a
a
a
a
"}
```

PowerfulBacon committed:
Added in ignoring space turfs option and ability to save weird shaped areas that aren't squares (null tiles are saved as NOOP)

PowerfulBacon commented:
Added in ability to ignore space turfs (Saves them as noop) as well as the ability to input weirdly shaped areas (if other coders want to save some stuff that isn't a square)
![image](https://user-images.githubusercontent.com/26465327/79565905-86588d00-80a9-11ea-93a0-b30c195df412.png)

PowerfulBacon committed:
Added the ability to choose what parts get saved as well as mob saving

PowerfulBacon committed:
Stops the saving of ghosts, as well as fixing inverted checks

PowerfulBacon committed:
Added in ability to save shuttle areas only

PowerfulBacon committed:
Removed redundant text as well as redundant comment

PowerfulBacon committed:
Adds in a custom proc to objects which will allow for individual objects to be given custom data to be added to the save file. For example, ORMs can have all of their sheets appear on the turf when saved so all the stuff inside them isn't lost because that would be sad. Oh and also the contents of things will now be stored on top of them (Lockers being emptied no more).

However, this will cause lockers that spawn with default things to duplicate their default items.

PowerfulBacon committed:
Adds in lockers saving their contents

PowerfulBacon committed:
oops

PowerfulBacon commented:
Implemented saving the contents of lockers
```
"cd" = (
/obj/structure/closet/lasertag/blue,
/obj/item/gun/energy/laser/bluetag,
/obj/item/gun/energy/laser/bluetag,
/obj/item/gun/energy/laser/bluetag,
/obj/item/clothing/suit/bluetag,
/obj/item/clothing/suit/bluetag,
/obj/item/clothing/suit/bluetag,
/obj/item/clothing/head/helmet/bluetaghelm,
/obj/machinery/light{
	dir = 1
	},
/turf/open/floor/plasteel,
/area/crew_quarters/fitness)
```
![image](https://user-images.githubusercontent.com/26465327/79574485-1e5d7300-80b8-11ea-965a-2d66aed2ea35.png)

PowerfulBacon committed:
Closets now save the data of data of things inside them

PowerfulBacon committed:
Depth check, to not crash server if somehow there is an infinite locker or something

PowerfulBacon committed:
Ore silos will now save their contents on top of the ore silo

PowerfulBacon commented:
Ore silos will now save their contents however, on loading the map the contents will spawn on top of the silo and will need to be placed back in. Better than nothing
![image](https://user-images.githubusercontent.com/26465327/79578404-17d1fa00-80be-11ea-8455-2112ea090455.png)

```
"p" = (
/obj/machinery/ore_silo,
/obj/item/stack/sheet/mineral/copper{
	amount = 42
	},
/obj/item/stack/sheet/mineral/silver{
	amount = 50
	},
/obj/item/stack/sheet/mineral/silver{
	amount = 13
	},
/obj/item/stack/sheet/mineral/gold{
	amount = 45
	},
/obj/item/stack/sheet/mineral/plasma{
	amount = 42
	},
/obj/item/stack/sheet/mineral/titanium{
	amount = 50
	},
/obj/item/stack/sheet/mineral/titanium{
	amount = 50
	},
/obj/item/stack/sheet/mineral/titanium{
	amount = 50
	},
/obj/item/stack/sheet/mineral/titanium{
	amount = 13
	},
/turf/open/floor/plasteel/dark,
/area/ai_monitored/nuke_storage)
```

PowerfulBacon commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -0,0 +1,37 @@
+	usr << ftp(filedir)
+
+	//Step 4: Remove the file from the server (hopefully we can find a way to avoid step)
+	fdel(filedir)
```
This doesn't actually delete the file from the server, we should remove it once we don't need it anymore

ike709 replied:
🤔

PowerfulBacon replied:
I really dunno what to do about this one, I believe it is because the file cannot be deleted while in ftp, maybe if I add a timer or something

PowerfulBacon committed:
Adds logging as well as admin check since this is saving stuff to server

PowerfulBacon commented:
This seems to be working now, needs review from someone since I am good at missing out on obvious things, but it seems to function decently.
Saved some pretty large areas
![image](https://user-images.githubusercontent.com/26465327/79583925-98482900-80c5-11ea-9c3b-aadacb996eaa.png)
This whole area and a little more was saved and loaded pretty nicely

PowerfulBacon commented:
Copy of the most recent file (converted to .txt)
[tgm_039 - Copy.txt](https://github.com/BeeStation/BeeStation-Hornet/files/4495007/tgm_039.-.Copy.txt)

PowerfulBacon committed:
Updates it so you can chagne the vars to save

PowerfulBacon committed:
Variables are now sanitised, otherwise you can name a drink "";},/obj/item/gun/energy/laser/instakill{name="da epic gun" and spawn an isntakill rifle

PowerfulBacon committed:
Only sanatises strings now

PowerfulBacon committed:
Changes sanatise characters so valid characters don't get sanatised

ike709 commented on code/modules/mapexporting/mapexporter.dm:
```diff
@@ -0,0 +1,169 @@
+	return allTurfs
+
+//vars_to_save = list() to save all vars
+/proc/generate_tgm_metadata(var/atom/O, var/list/vars_to_save = list("pixel_x", "pixel_y", "dir", "name", "req_access", "req_access_txt", "piping_layer", "color", "icon_state", "pipe_color", "amount"))
```
There should be an easy way to modify the `vars_to_save` on a per-type basis. Have a look at how World Server does it with their `vars_to_save()` proc (IIRC that's the name of it).

PowerfulBacon replied:
There is now an empty variable on every atom called vars_to_save, if the variable isn't set to anything the default vars to save will be saved (the one passed into the function), else the vars to save of that atom will be saved.

<s>There is also a flag that works on objects and mobs called DONT_SAVE_1</s>
Edit: I hate this idea, adding it to all objectives. I am just going to have a blacklist

ike709 commented:
There should also be a way to prevent certain things from being saved on a per-type basis, like the nuke disk.

PowerfulBacon committed:
Adds an obj black list and atom var 'vars_to_save'

qwertyquerty commented:
@ike709 are you at all concerned with adding another list to /atom?

PowerfulBacon commented:
> @ike709 are you at all concerned with adding another list to /atom?

Huh I thought I changed it to a return value, but I guess I was too lazy to push it

PowerfulBacon committed:
Replaces a stupid list on every atom with a return list

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -0,0 +1,51 @@
+		return
+	//Emergency check
+	if(L.map.len > 1600)
+		var/confirm = alert("Uhm, are you sure, the are is quiet large?", "Run generator", "Yes", "No")
```
```suggestion
		var/confirm = alert("Uhm, are you sure, the area is quiet large?", "Run generator", "Yes", "No")
```

ike709 commented on code/modules/mapexporting/mapexporter.dm:
```diff
@@ -0,0 +1,181 @@
+//This has been made semi-modular so you should be able to use these functions
+//elsewhere in code if you ever need to get a file in the .dmm format
+/atom/proc/get_save_vars()
+	return list()
```
```suggestion
	return
```

PowerfulBacon committed:
Update code/modules/mapexporting/mapexporter.dm

Co-authored-by: ike709 <ike709@users.noreply.github.com>

PowerfulBacon committed:
Update code/modules/buildmode/submodes/save_area.dm

Co-authored-by: ike709 <ike709@users.noreply.github.com>

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -0,0 +1,51 @@
+		if(confirm != "Yes")
+			return
+	//Log just in case something happens
+	log_game("[ckey(usr)] ran the save level map generator on [L.map.len] turfs.")
```
Should also message admins here.

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -0,0 +1,51 @@
+
+	//Step 1: Get the data (This can take a while)
+	var/dat = convert_map_to_tgm(L.map)
+
```
Somewhere around here you should send a confirmation message to the person that it's doing stuff.

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -0,0 +1,51 @@
+
+	//Step 3: Give the file to client for download
+	usr << ftp(filedir)
+
```
And finally a message to the user that it has completed.

ike709 reviewed (changes requested):
It seems to work but you need to improve user feedback, see other comments.

ike709 commented:
Also I don't see what the point of doing all the work for TGM was if you're not going to add the comment to the top of the file that says it's TGM, which is required for it to actually count as TGM.

ike709 commented:
Also it doesn't save correctly, so DNM pending that getting fixed. Check discord for details.

PowerfulBacon committed:
Fixes me being stupid and also adds the TGM comment

PowerfulBacon committed:
Ike changes

PowerfulBacon committed:
Changes position of check so you can't run 2 at the same time

PowerfulBacon committed:
lol

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -31,8 +36,13 @@
 		var/confirm = alert("Uhm, are you sure, the area is quiet large?", "Run generator", "Yes", "No")
```
You need to check `GLOB.save_area_executing` again here in case they were sitting on the alert popup for a while and someone else started a save.

PowerfulBacon replied:
oh I thought I moved it, must have forgot to push

ike709 replied:
You did, I'm dumb and reviewed the commit instead of all the changes.

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -31,8 +36,13 @@
+	GLOB.save_area_executing = TRUE
+
 	//Log just in case something happens
 	log_game("[ckey(usr)] ran the save level map generator on [L.map.len] turfs.")
```
```suggestion
	log_game("[key_name(usr)] ran the save level map generator on [L.map.len] turfs.")
```

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -31,8 +36,13 @@
+
 	//Log just in case something happens
 	log_game("[ckey(usr)] ran the save level map generator on [L.map.len] turfs.")
+	message_admins("[ckey(usr)] ran the save level map generator on [L.map.len] turfs.")
```
```suggestion
	message_admins("[key_name(usr)] ran the save level map generator on [L.map.len] turfs.")
```

PowerfulBacon committed:
Update code/modules/buildmode/submodes/save_area.dm

Co-authored-by: ike709 <ike709@users.noreply.github.com>

PowerfulBacon committed:
Update code/modules/buildmode/submodes/save_area.dm

Co-authored-by: ike709 <ike709@users.noreply.github.com>

ike709 commented on code/modules/buildmode/submodes/save_area.dm:
```diff
@@ -50,3 +60,5 @@
 	fdel(filedir)
 	log_game("[L.map.len] turfs have been saved by [ckey(usr)]")
+	alert("Area saved successfully.", "Action Successful!", "Ok")
+	save_area_executing.save_area_executing = FALSE
```
```suggestion
	GLOB.save_area_executing = FALSE
```

ike709 commented:
I did a dumb and reviewed one of the commits instead of all of the recent changes. This is fine. Will testmerge again later.
